src/database-controller/src/core/framework.js

```diff
@@ -0,0 +1,421 @@
+          );
+        } else {
+          throw err;
+        }
+      }
+    }
+    if (this._dockerSecretDef) {
+      try {
+        await k8s.createSecret(this._dockerSecretDef);
+      } catch (err) {
+        if (err.response && err.response.statusCode === 409) {
+          logger.warn(
+            `Secret ${this._dockerSecretDef.metadata.name} already exists.`,
+          );
+        } else {
+          throw err;
+        }
+      }
+    }
+  }
+
+  silentPatch(frameworkResponse) {
+    // do not await for patch
+    this._configSecretDef &&
+      k8s
+        .patchSecretOwnerToFramework(this._configSecretDef, frameworkResponse)
+        .catch(ignoreError);
+    this._dockerSecretDef &&
+      k8s
+        .patchSecretOwnerToFramework(this._dockerSecretDef, frameworkResponse)
+        .catch(ignoreError);
+  }
+
+  silentDelete() {
+    // do not await for delete
+    this._configSecretDef &&
+      k8s.deleteSecret(this._configSecretDef.metadata.name).catch(ignoreError);
+    this._priorityClassDef &&
+      k8s
+        .deletePriorityClass(this._priorityClassDef.metadata.name)
+        .catch(ignoreError);
+    this._dockerSecretDef &&
+      k8s.deleteSecret(this._dockerSecretDef.metadata.name).catch(ignoreError);
+  }
+
+  getUpdate() {
+    const update = {};
+    if (this._configSecretDef) {
+      update.configSecretDef = JSON.stringify(this._configSecretDef);
+    }
+    if (this._priorityClassDef) {
+      update.priorityClassDef = JSON.stringify(this._priorityClassDef);
+    }
+    if (this._dockerSecretDef) {
+      update.dockerSecretDef = JSON.stringify(this._dockerSecretDef);
+    }
+    return update;
+  }
+}
+
+async function synchronizeCreate(snapshot, addOns) {
+  await addOns.create();
+  try {
+    const response = await k8s.createFramework(snapshot.getRequest(false));
+    // framework is created successfully.
+    const frameworkResponse = response.body;
+    // don't wait for patching
+    addOns.silentPatch(frameworkResponse);
+    return frameworkResponse;
+  } catch (err) {
+    if (err.response && err.response.statusCode === 409) {
+      // doesn't delete add-ons if 409 error
+      logger.warn(`Framework ${snapshot.getName()} already exists.`);
+      throw err;
+    } else {
+      // delete add-ons if 409 error
+      addOns.silentDelete();
+      throw err;
+    }
+  }
+}
+
+async function synchronizeModify(snapshot) {
+  const response = await k8s.patchFramework(
+    snapshot.getName(),
+    snapshot.getRequest(false),
+  );
+  const frameworkResponse = response.body;
+  return frameworkResponse;
+}
+
+async function synchronizeRequest(snapshot, addOns) {
+  // any error will be raised
+  // if succeed, return framework from api server
+  // There may be multiple calls of synchronizeRequest.
+  // Poller and write-merger uses this method.
+  try {
+    await k8s.getFramework(snapshot.getName());
+    // if framework exists
```

Comment:
Can skip this call

Reply:
Instead of error handling, I prefer this way to make the code more clear.

Reply:
This will cause k8s unnecessary call, esp. when the framework is very large, so remove getFramework call is more important.

Reply:
Pls check this

---
In reply to: [466813269](https://github.com/microsoft/pai/pull/4761#discussion_r466813269) [](ancestors = 466813269)